src/charm.py

```diff
@@ -135,33 +142,8 @@ def _on_grafana_changed(self, event):
             self.model.get_binding(event.relation).network.bind_address
         )
 
-    def _on_alertmanager_changed(self, event):
-        """Set an alertmanager configuration.
-
-        In response to any changes in relations with Alertmanager,
-        the list of currently available Alertmanagers is updated,
-        a new Prometheus configuration set and Prometheus is
-        restarted.
-        """
-        if not self.unit.is_leader():
-            return
-
-        addrs = json.loads(event.relation.data[event.app].get('addrs', '[]'))
-
-        self._stored.alertmanagers = addrs
-
-        self._configure()
-
-    def _on_alertmanager_broken(self, event):
-        """Remove all alertmanager configuration.
-
-        When an Alertmanager departs it is removed from the list
-        of currently available Alertmanagers, the Prometheus configuration
-        is updated and Prometheus is restarted.
-        """
-        if not self.unit.is_leader():
-            return
-        self._stored.alertmanagers.clear()
+    def _on_alertmanager_cluster_changed(self, event):
+        self._stored.alertmanagers = self.alertmanager_lib.get_cluster_info()
```

Comment:
Is there any reason to store the list of alert managers both in the AlertmanagerConsumer library and the Prometheus Charm ? Why not just integrate it into `_alerting_config()` ?

Reply:
I intentionally kept changes to the prom op to a minimum, to focus on the interface only.

Reply:
Can I request you to create a ticket for future refactor either in prometheus repo or alertmanager repo or both, if you agree that a refactor here may be a good idea. I think it does not make sense to store it in two places. Even more it does not make sense to store it at all. You already have a ticket to that effect (i.e. to use app data).

Reply:
removed `_stored.alertmanagers`